[SCEV] Add initial support for ptrtoaddr.

fhahn opened the pull request:
Add initial support for PtrToAddr to SCEV, including a new SCEVPtrToAddrExpr and SCEV expansion support for it.

nikic reviewed (commented):
Happy to have this support, but I don't think we should use it for trip count calculation at this point, and only for explicit `ptrtoaddr` in IR. We have zero optimization support for this instruction currently, so I don't think we should be emitting it yet.

fhahn committed:
[SCEV] Add initial support for ptrtoaddr.

Add support for PtrToAddr to SCEV and use it to compute pointer difference when computing backedge taken counts.

This patch retains SCEVPtrToInt, which is mostly used to expressions comping directly from IR.

PtrToAddr more closely matches the semantics of most uses in SCEV. See #156978 for some related discussion.

The patch still retains SCEVPtrToInt, as it is used for cases where a SCEV expression is used in IR by a ptrtoint, e.g. https://github.com/llvm/llvm-project/blob/main/llvm/test/Transforms/IndVarSimplify/pr59633.ll. Some (or perhaps all, still need to check) should probably just be SCEVUnknown.

fhahn commented:
The latest version should just have the bits needed to support ptrtoaddr LLVM IR instructions for now

fhahn commented:
(Still need to update polly...)

nikic commented on llvm/test/Analysis/ScalarEvolution/ptrtoaddr.ll:
We shouldn't test `-disable-verify` here. Unlike ptrtoint, the ptrtoaddr result type is fixed (and equal to the "effective SCEV type" of the pointer), so any zext/trunc handling is not applicable for this instruction. `getPtrToAddrExpr()` should probably not even accept a type argument?

fhahn replied:
Yeah, I removed the pointer argument and asserted that the conversion is always lossless, thanks

fhahn committed:
!fixup strip type argument

fhahn committed:
!fixup fix polly build again

fhahn committed:
!fixup add 32 bit index width test, remove left-over check lines

fhahn commented on llvm/test/Analysis/ScalarEvolution/ptrtoaddr.ll:
```diff
@@ -1,78 +1,77 @@
+; RUN: opt -passes='print<scalar-evolution>' -disable-output %s 2>&1 | FileCheck %s
 
-declare void @useptr(ptr)
+target datalayout="e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
```
Yep, I added it as a seperate test file, thanks

fhahn commented on llvm/test/Analysis/ScalarEvolution/ptrtoaddr.ll:
```diff
@@ -1,78 +1,77 @@
+; X32-NEXT:    --> (trunc i32 (ptrtoaddr ptr %in to i32) to i16) U: full-set S: full-set
 ; X32-NEXT:    %p3 = ptrtoaddr ptr %in to i128
-; X32-NEXT:    --> %p3 U: full-set S: full-set
+; X32-NEXT:    --> (zext i32 (ptrtoaddr ptr %in to i32) to i128) U: [0,4294967296) S: [0,4294967296)
```
Dropped, thanks

fhahn committed:
!fixup fix i32 index width tests, bail out

nikic commented on llvm/test/Analysis/ScalarEvolution/ptrtoaddr-i32-index-width.ll:
```diff
@@ -0,0 +1,70 @@
+; CHECK-LABEL: 'ptrtoaddr'
+; CHECK-NEXT:  Classifying expressions for: @ptrtoaddr
+; CHECK-NEXT:    %p0 = ptrtoaddr ptr %in to i32
+; CHECK-NEXT:    --> %p0 U: full-set S: full-set
```
For ptrtoaddr we *do* want to create SCEV expressions for these. Handling the case where the address size doesn't match the pointer size is important.

I think the only case we might want to reject for ptrtoaddr is the one with hasUnstableRepresentation.

I feel like for ptrtoaddr the whole distinction of "lossless ptrtoaddr" doesn't really make sense (given that the result width is fixed). I think we should have only getPtrToAddr(). We should share only the core implementation of getLosslessPtrToIntExpr, without the data layout checks.

nikic replied:
With the clarification to icmp semantics from https://github.com/llvm/llvm-project/pull/163936, it is also fine to use ptrtoaddr with mismatched pointer/address size in all the places that use getLosslessPtrToIntExpr() on icmps (like computeExitLimit and helpers).

fhahn committed:
!fixup don't use getLosslessPtrToIntExpr for PtrToAddr

fhahn committed:
!fixup fix formatting

fhahn committed:
!fixup fix polly build failure

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -1131,6 +1143,40 @@ const SCEV *ScalarEvolution::getLosslessPtrToIntExpr(const SCEV *Op,
+  // If not, is this expression something we can't reduce any further?
+  if (auto *U = dyn_cast<SCEVUnknown>(Op)) {
+    // Perform some basic constant folding. If the operand of the ptr2int cast
+    // is a null pointer, don't create a ptr2int SCEV expression (that will be
```
```suggestion
    // Perform some basic constant folding. If the operand of the ptr2addr cast
    // is a null pointer, don't create a ptr2addr SCEV expression (that will be
```

nikic reviewed (approved):
LGTM

fhahn committed:
!fixup remove type argument, fix comment, thanks

nikic reviewed (approved):
LGTM

fhahn committed:
Merge remote-tracking branch 'origin/main' into scev-ptrtoaddr

fhahn committed:
!fixup lex order.
